presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushTopNIntoTableScan.java

```diff
@@ -56,33 +55,30 @@ public PushTopNIntoTableScan(Metadata metadata)
     }
 
     @Override
-    public Result apply(TopNNode node, Captures captures, Context context)
+    public Result apply(TopNNode topNNode, Captures captures, Context context)
     {
         TableScanNode tableScan = captures.get(TABLE_SCAN);
 
-        long topNCount = node.getCount();
-        List<SortItem> sortItems = node.getOrderingScheme().toSortItems();
+        long topNCount = topNNode.getCount();
+        List<SortItem> sortItems = topNNode.getOrderingScheme().toSortItems();
 
         Map<String, ColumnHandle> assignments = tableScan.getAssignments()
                 .entrySet().stream()
                 .collect(toImmutableMap(entry -> entry.getKey().getName(), Map.Entry::getValue));
 
-        Optional<TableHandle> result = metadata.applyTopN(
-                context.getSession(),
-                tableScan.getTable(),
-                topNCount,
-                sortItems,
-                assignments);
+        return metadata.applyTopN(context.getSession(), tableScan.getTable(), topNCount, sortItems, assignments)
+                .map(result -> {
+                    PlanNode node = TableScanNode.newInstance(
+                            context.getIdAllocator().getNextId(),
+                            result.getHandle(),
+                            tableScan.getOutputSymbols(),
+                            tableScan.getAssignments());
 
-        if (result.isEmpty()) {
-            return Result.empty();
-        }
-
-        return Result.ofPlanNode(
-                        TableScanNode.newInstance(
-                                context.getIdAllocator().getNextId(),
-                                result.get(),
-                                tableScan.getOutputSymbols(),
-                                tableScan.getAssignments()));
+                    if (!result.isTopNGuaranteed()) {
+                        node = new TopNNode(topNNode.getId(), node, topNNode.getCount(), topNNode.getOrderingScheme(), TopNNode.Step.PARTIAL);
```

Comment:
I believe this should be `Step.FINAL`. The partial Top N is being handled by the connector at this point.

Reply:
Fixed.

Reply:
Followed up here: https://github.com/trinodb/trino/pull/4249#discussion_r580980702